### PR TITLE
First pass on install-related Client Hints

### PR DESCRIPTION
**Pull request by aarongustafson**

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -411,6 +411,76 @@ <h3>
...
+        order to evaluate the success of marketing efforts within a given
+        catalog. In the traditional software distribution world, app catalogs
+        enable developers to see how many users installed their app from that
+        catalog. With web-based distribution, this sort of telemetry is
+        impossible to obtain. To fill this gap, this spec provides for the
+        following <a data-cite=
+        "client-hints-infrastructure#client-hint-token">client hint token</a>s:
+      </p>
+      <ul>
+        <li>`Sec-CH-App-Installed` whose value is a suitable [=Installation
```
Comment (marcoscaceres) [inline index.html:429]: I'm a bit worried about conflating the two things. In particular, I'm not entirely sure why "installed from" would be sent as  a client hint (and be sent on every request, at that)? It feels like "installed from" would be established once as, say, a `Referer:` header or similar (or via some kind of redirect from the store, where such analytics information would be stored). 

Unlike the other CH, it's unclear that this plays a "critical" role in the presentation/adaption of the web application. So, IMO, this doesn't seem to be a good fit for a CH. 

Additionally, unlike other CHs, the "Installation Source value" could be used as a super cookie, as it's would be set by the store. For example: `Sec-CH-App-Installation-Source: https://store.com?userid=123` would allow respawning of cookies, allow for tracking, etc.

File: index.html [hunk 2]
```diff
@@ -411,6 +411,76 @@ <h3>
...
+        It is possible that these headers <a data-cite=
+        "https://wicg.github.io/ua-client-hints/#fingerprinting">may be used
+        for passive fingerprinting of a user</a> and User Agents should take
+        care to <a data-cite=
+        "https://wicg.github.io/ua-client-hints/#access">restrict access</a> to
+        this information.
+      </p>
+      <section>
+        <h3>
+          The 'Sec-CH-App-Installed' Header Field
```
Comment (marcoscaceres) [inline index.html:446]: As described in the other manifest bug, what was "critical" was to know ahead of time what the display mode the application is in, rather than if it's installed or not. It's foreseeable that an app could be run in some display mode declared in the manifest without it having been "installed" (which is one use case for Web Manifest, but not its exclusive usage). By changing this to, say, `'Sec-CH-Display-Mode'` with values `"browser" | "fullscreen" | "standalone"` etc. instead, it also gets rid of the fingerprinting aspect (doesn't make it any worse), while also being more privacy sensitive: "the app happens to be running in some mode, as exposed already via CSS, but installed aspect remains uncertain".

**Review by marcoscaceres — COMMENTED** [review]
Thanks for drafting this together @aarongustafson! some thoughts...

File: index.html (hunk 2 above)
Comment (marcoscaceres) [reply index.html:446]: Just as an aside, as this hint affects browsers, we might want to eventually move it over to the main spec.

Comment (dmurph) [reply index.html:446]: I also prefer display mode, as I think this is more directly what the prerendering use case needs. Also, since web apps can still have different display modes (minimal-ui vs standalone), I think developers would still need the display mode.

For example, for a site that needs navigation controls, if the display mode is minimal-ui, then the UI would not need back/forward buttons, but if it is standalone, then the UI would need back/forward buttons.

It is also a nicely direct parallel to the existing display mode css query.

Comment (dmurph) [reply index.html:446]: On the other hand, if you have use cases that support doing it this way, definitely let me know!

File: index.html (hunk 1 above)
Comment (aarongustafson) [reply index.html:429]: > I'm a bit worried about conflating the two things. In particular, I'm not entirely sure why "installed from" would be sent as a client hint (and be sent on every request, at that)? It feels like "installed from" would be established once as, say, a Referer: header or similar (or via some kind of redirect from the store, where such analytics information would be stored).
>
> Unlike the other CH, it's unclear that this plays a "critical" role in the presentation/adaption of the web application. So, IMO, this doesn't seem to be a good fit for a CH.

Interestingly `Referer` was my first choice, and at launch in the absence of an inbound `Referer` value, but some privacy folks I was discussing this with said that `Referer` didn’t seem relevant and that it was getting locked down. They suggested that I look at Client Hints…

Personally, I think `Referer` makes the most sense for this information as well.

> Additionally, unlike other CHs, the "Installation Source value" could be used as a super cookie, as it's would be set by the store. For example: Sec-CH-App-Installation-Source: https://store.com?userid=123 would allow respawning of cookies, allow for tracking, etc.

No. As outlined here, the reported value would be a catalog-type string value outlined in the `platform` key. There would be no opportunity for additional information to be added. That’s not to say that fingerprinting could not be enabled by this value, but the header itself would not divulge user-specific information. My expectation is that any implementation of a "privacy budget" or similar would consider this information as increasing to the overall fingerprinting surface.

File: index.html (hunk 2 above)
Comment (aarongustafson) [reply index.html:446]: @marcoscaceres @dmurph I should have been more clear when I linked to this from that discussion: I was planning to take the feedback on this over to the Manifest and implement this feature there as `Sec-CH-App-Display-Mode`. I’ve not written language defining a Client Hint within another spec before and thought this would be a cleaner place to do that initial learning.

Comment (marcoscaceres) [reply index.html:446]: Totally fine if you want to stage/incubate it here. I have no preference.

Comment (aarongustafson) [reply index.html:446]: Yeah, I mainly wanted to test out the section organization & such here. Then I will use that as a reference for the Manifest PR. I’d be happy to move the "installed" CH there as well. I think having an explicit flag for "this is installed" would be better than having to infer it from the display hint (or worse, the CSS to JS bridge). This would enable the server side to, for instance, avoid shipping code that would prompt installation.

Comment (marcoscaceres) [reply index.html:446]: > This would enable the server side to, for instance, avoid shipping code that would prompt installation.

This harps back to this being a UI issues for UAs that do somewhat intrusive install prompting, and why both Mozilla and Apple rejected `BeforeInstallPrompt` (and "install" generally), and instead used "ambient badging" and/or having "add to homescreen" as a menu option. 

I'm worried that the "installed" CH reinforces the whole "installed / not installed" dichotomy, which doesn't fit well with the whole idea of PWAs being a progressive enhancement: that is, a web application should work exactly the same if it's installed or not. It MAY differ when displayed differently (`CH-Display-Mode` and `'display-mode'` media feature), but identifying it as "installed" should make no difference and could be an anti-pattern.

Comment (aarongustafson) [reply index.html:446]: > identifying it as "installed" should make no difference and could be an anti-pattern.

I agree… to an extent. I think there are instances where installation will make a difference. For example, something like the ability to run on OS login is something that should likely be reserved for installed PWAs. Similarly, I know many of us have discussed relaxing storage limits for installed PWAs. Additionally, I think the File System access spec has different behavior with respect to permissions prompting when you are installed vs. not.

To be clear, I think the places where the experience of tab vs. installed should be as few as possible, but at the same time I acknowledge that there are exceptions that we should take into account.

> This harps back to this being a UI issues for UAs that do somewhat intrusive install prompting, and why both Mozilla and Apple rejected BeforeInstallPrompt (and "install" generally), and instead used "ambient badging" and/or having "add to homescreen" as a menu option.

I wasn’t even talking about `beforeInstallPrompt`, specifically. I was thinking about some of the other ways website advertise installation on their web pages, whether that’s somethign simple like a button advertising "download in the X store" or custom controls like [PWA Builder’s rich installer web component](https://components.pwabuilder.com/component/install_pwa). These would be confusing to render and extraneous to deliver if the PWA is already installed.

Comment (marcoscaceres) [reply index.html:446]: > I agree… to an extent. I think there are instances where installation will make a difference. For example, something like the ability to run on OS login is something that should likely be reserved for installed PWAs.

Agree, but being run at startup is not detectable by the application itself. That would be user configurable at the OS level, like on MacOS:

Screen Shot 2021-03-24 at 2 40 32 pm

I imagine there are other equivalents in other OSs. So the hint doesn't help here as no assumption can be made. 

> Similarly, I know many of us have discussed relaxing storage limits for installed PWAs. 

Again, agree, but that is not observable by the app and not something the application can assume, as all storage is either artificially or physically limited. Such a hint wouldn't help here, either.

> Additionally, I think the File System access spec has different behavior with respect to permissions prompting when you are installed vs. not.

But again, an implementation detail built into the API itself - but won't be observable by the application/app-developer. So again, the hint doesn't really help, except in potentially allowing the developer to make bad assumptions.

You see where I'm going with this, right? The hint ends up being a footgun because any assumptions made based on being installed turn out to be harmful to either the app or the end user. 

> I wasn’t even talking about beforeInstallPrompt, specifically.

I'm sorry, I know. They are somewhat interconnected - bear with me. 

> I was thinking about some of the other ways website advertise installation on their web pages, whether that’s something simple like a button advertising "download in the X store" 

Ah, I think this is our point of departure! 

Consider: If, as a user, I'm in the web application, and I'm happily using the web application, then why would I navigate all the way to another app store to install the application I'm already using? If I want to "install it", I just hit "add to homescreen" right there in the UI without the "app store" middle-person. 

IIUC, what you are describing is a different user story: as a user, I'm currently not using the "App X", but I'm searching for a cool app in the App Store. The store recommends "App X", which I'm then navigate to "add to home screen" or "install". 

Taken further, if the App Store starts wrapping the web app in an APK, or starts doing other funky things (e.g., enabling proprietary APIs only for apps installed from the a particular store), then all bets are off. 

> or custom controls like PWA Builder’s rich installer web component. These would be confusing to render and extraneous to deliver if the PWA is already installed.

Correct, which is generally not what we want (apps showing install prompts). The point of the manifest was for browsers to do the prompting on behalf of websites.

Not Mozilla specific, but there is a detailed discussion on this subject:
https://github.com/mozilla/standards-positions/issues/84#issue-314879910

And more discussion, which lead to further removal of the whole idea of "installed" from Web Manifest:
https://github.com/w3c/manifest/issues/627

**Comment by dmurph** [comment]
I think a good road forward is for us to gather partners with explicit use cases (or even internal customers with explicit use cases) where these client hints are required. I know of one major one currently:

* server-side rendering is blocked for webapps where the display mode influences the content - e.g. a fullscreen view is different than a standalone, different than minimal-ui, different than tabbed (browser), etc. An easy example is minimal-ui vs standalone, and the need to present navigation controls if in standalone

and one minor use case that could greatly increase adoption:

* established properties use server-side logging frameworks for metrics and experiments, and it is really difficult to pipe back data from javascript on the client to join with the server side logs to know 'is this user viewing the site as a webapp?'. Providing this information as a client hints unblocks these sites from putting major resources towards webapps because they can now run experiments to determine things like "relation between user satisfaction / retention and web app usage (/ display mode)."

I think both of these are solved by having the 'display mode' client hint. 


I think I generally agree with @marcoscaceres  that we don't want to encourage the case for the install prompt, and if we are getting that feedback I really want to dig in and know why that partner wants the webapp to be installed.
I think the one use case you are citing

File: index.html (hunk 2 above)
Comment (aarongustafson) [reply index.html:446]: > You see where I'm going with this, right? The hint ends up being a footgun because any assumptions made based on being installed turn out to be harmful to either the app or the end user.

Agree on making assumptions, devs should be cautioned against conflating installation with anything other than being an indication that the app is installed. The reason I was thinking the Client Hint (or `Referer` header) made sense is that the site content might vary based on knowing that installation had happened. For instance:

* If a user is using the web app in a tab, the content could adjust to let them know that a particular feature is not available to them (e.g., run on OS startup) or that the behavior of a certain crucial API might behave differently (e.g., File System access requiring re-authorization to access system files in a tab rather than having long-lived permission in an installed app).
* Only offering offline media storage of low/mid quality when not installed because you know you’re abiding by the regular storage pool limits.
* Removing references to browser tabs in the content when the web app is installed. (No UA has tabbed PWAs yet, to my knowledge.)

> Consider: If, as a user, I'm in the web application, and I'm happily using the web application, then why would I navigate all the way to another app store to install the application I'm already using? If I want to "install it", I just hit "add to homescreen" right there in the UI without the "app store" middle-person.

I agree with you on most of this. I’m not suggesting folks need to go through an app store at all — though I do see the appeal for some products as a way to increase their discoverability and increase their market share. That said, browser install-ability, while highly desirable, still suffers from discoverability issues. That’s why some folks opt to include installation prompts like the widget PWA Builder created. If the web app is already installed, there would be no need to ship that content/code to the client (and potentially confuse their users).

Additionally, there are web apps that have native client equivalents that they link to from within their app (Basecamp, for example). If they had some way to know the web app was already installed, they could remove those advertisements altogether.

Screen Shot 2021-04-05 at 5 10 35 PM

> > or custom controls like PWA Builder’s rich installer web component. These would be confusing to render and extraneous to deliver if the PWA is already installed.
> 
> Correct, which is generally not what we want (apps showing install prompts). This is bad practice:
> 
> Example install prompts

Those are horrible, yes. I think [PWA Builder’s component](https://pwainstall.glitch.me/) is a little more elegant.

> The point of the manifest was for browsers to do the prompting on behalf of websites.

Agreed, but browser’s have not figured out the best/most understandable/easily discoverable UI yet, so in the meantime…

> And more discussion, which lead to further removal of the whole idea of "installed" from Web Manifest:
> [w3c/manifest#627](https://github.com/w3c/manifest/issues/627)

Thanks for that.

When we boil it all down, I think there is a case for informing a site that it is currently installed. Folks are already inferring this from the `display-mode` in CSS (and would do the same with the `dlsplay_mode` Client Hint, I reckon). Such assumptions are problematic, as you note; for example, consider that using the `matchMedia` path to determine whether you are installed gives a false positive when the Fullscreen API is invoked (where `window.matchMedia( '(display-mode: fullscreen)' ).matches` is `true`). My hope is that the prospective Client Hint would specifically indicate that this false-positive should not happen, but it still operates on the same basic premise that a display mode being applied indicates the app has been installed, which may not be the case.

I’d rather be explicit that this app is installed or it isn’t.

**Comment by aarongustafson** [comment]
> I think a good road forward is for us to gather partners with explicit use cases (or even internal customers with explicit use cases) where these client hints are required. 

Agreed.

> * server-side rendering is blocked for webapps where the display mode influences the content - e.g. a fullscreen view is different than a standalone, different than minimal-ui, different than tabbed (browser), etc. An easy example is minimal-ui vs standalone, and the need to present navigation controls if in standalone

Yes. This is an excellent case for the display mode hint, which I will work on shortly, over in the Manifest spec. Setting aside the topics of this PR for a moment, how are y’all feeling about the organization of the Client Hints section? Do you feel it could be replicated in the manifest for display mode?

> and one minor use case that could greatly increase adoption:
> 
> * established properties use server-side logging frameworks for metrics and experiments, and it is really difficult to pipe back data from javascript on the client to join with the server side logs to know 'is this user viewing the site as a webapp?'. Providing this information as a client hints unblocks these sites from putting major resources towards webapps because they can now run experiments to determine things like "relation between user satisfaction / retention and web app usage (/ display mode)."

100% agree on this. Having access to this info without JavaScript would be huge for analytics on static sites as well.

> I think both of these are solved by having the 'display mode' client hint.

I am not on board with equating display mode with installation unless we decide to specifically make that part of the implementation requirements in the spec for the display mode Client Hint. The current equivalent approach using `matchMedia()` suffers from false positives (see [my comment above](https://github.com/w3c/manifest-app-info/pull/32#discussion_r607411613)), so I don’t want to replicate that issue, nor do I want to create a situation where devs make assumptions about a web app’s installation status based on some other factor; I’d rather just explicitly tell them they’re installed.

> I think I generally agree with @marcoscaceres that we don't want to encourage the case for the install prompt, and if we are getting that feedback I really want to dig in and know why that partner wants the webapp to be installed.

To be clear, I’m not suggesting web apps be able to prompt for their own installation (or that they shouldn’t), I was merely bringing up that some content may vary based on whether the app is installed (or not). An advertisement to install the app was an example of that.

With that in mind, my third use case would be:

> Web apps are currently forced to deliver the same content and scripts to all users and cannot vary those based on whether the app is installed. For example, they may choose not to include the content and images advertising the availability of their app in an app catalog if they know it is already installed on this device.

File: index.html (hunk 2 above)
Comment (marcoscaceres) [reply index.html:446]: > Additionally, there are web apps that have native client equivalents that they link to from within their app (Basecamp, for example). If they had some way to know the web app was already installed, they could remove those advertisements altogether.

Native app counterpart to a web app generally serve different purposes (or provide entirely different experiences). Thus, in most cases, the installation of one does not preclude the need to install the other - so I don't think we will see those advertisements vanishing, as much as we strive for parity with native apps. 

> I’d rather be explicit that this app is installed or it isn’t.

We've illustrated both sides of the augment, which has been really fruitful and thought provoking. I see a few encouraging ❤️ from @hober, so I'm wondering if we should invite a few more folks to review this proposal and comment to get some fresh perspectives.

Comment (aarongustafson) [reply index.html:446]: Sounds good to me @marcoscaceres.

In [my comment replying to @dmurph](https://github.com/w3c/manifest-app-info/pull/32#issuecomment-813745047), below, I asked the following:

> Setting aside the topics of this PR for a moment, how are y’all feeling about the organization of the Client Hints section? Do you feel it could be replicated in the manifest for display mode?

Are you feeling that the organization of this section makes sense? If so, I can draft a PR for the Manifest on the less controversial Display Mode Client Hint.

Comment (dmurph) [reply index.html:446]: The structure seems OK to me, @yoavweiss do you have any thoughts on how specs should be adding client hints? Should it basically just model the hints here:
https://wicg.github.io/ua-client-hints/#http-ua-hints

Comment (dmurph) [reply index.html:446]: Also, @b1tr0t for FYI on this discussion

File: index.html [hunk 3]
```diff
@@ -411,6 +411,76 @@ <h3>
...
+          field gives a server information about whether or not a user agent is
+          representing the app in an "installed" state. It is a <a data-cite=
+          "rfc8941#section-3">Structured Header</a> whose value MUST be a
+          <a data-cite="rfc8941#section-3.3.6">boolean</a>.
+        </p>
+        <p>
+          The header's ABNF is:
+        </p>
+        <pre>
+          Sec-CH-App-Installed = sh-boolean
```
Comment (yoavweiss) [inline index.html:460]: That should be `sf-boolean`.
We should fix UA-CH as well...

File: index.html [hunk 4]
```diff
@@ -411,6 +411,76 @@ <h3>
...
+          triggered installation of the app. It is a <a data-cite=
+          "rfc8941#section-3">Structured Header</a> whose value MUST be a
+          <a data-cite="rfc8941#section-3.3.3">string</a>. Its value SHOULD
+          match one of the [=application distribution platforms=].
+        </p>
+        <p>
+          The header's ABNF is:
+        </p>
+        <pre>
+          Sec-CH-App-Installation-Source = sh-string
```
Comment (yoavweiss) [inline index.html:480]: `sf-string`

File: index.html (hunk 2 above)
Comment (yoavweiss) [reply index.html:446]: Commented on the structure. Overall LG.

A couple of comments in general:
* Client Hints [must not ](https://datatracker.ietf.org/doc/rfc8942/?include_text=1#:~:text=Therefore%2C%20features%20relying%20on%20this%20document%20to%20define%20Client%20Hint%20%20%20%20headers%20MUST%20NOT%20provide%20new%20information%20that%20is%20otherwise%20not%20made%20%20%20%20available%20to%20the%20application%20by%20the%20user%20agent%2C%20such%20as%20existing%20%20%20%20request%20headers%2C%20HTML%2C%20CSS%2C%20or%20JavaScript.) expose information that's not available through JS, CSS or HTML. Display mode is available. Installed state and origin is not AFAIK.
* Client Hints are meant for content negotiation. As @marcoscaceres pointed out, the use cases for display mode for content negotiation are clearer than the ones for an "installed" signal. And I don't think I heard content-negotiation reasons for including an "install origin" one. I understand the analytics appeal, but it seems like something that can be done with JS without requiring a Client Hint.

**Comment by aarongustafson** [comment]
@marcoscaceres @dmurph @yoavweiss I made some tweaks to move the installation source to using `Referer` (which is, incidentally, what Google Play TWAs do now, as I understand it—and, FWIW, was my initial choice for this info). I also tweaked some of the language around the "installed" hint (which I have heard is something devs would like to have). Since CHs require a JS equivalent, what do folks think about proposing `navigator.installed` (or `navigator.isInstalled`) as a boolean?

**Comment by marcoscaceres** [comment]
>  Since CHs require a JS equivalent, what do folks think about proposing navigator.installed (or navigator.isInstalled) as a boolean?

I'd be personally opposed to that for the reasons I gave. However, if others are supportive, I'd go with the majority. 

@annevk maybe you can give a Gecko perspective here? Not sure who to ping to get a Mozilla position on things like this.

**Comment by annevk** [comment]
We're not convinced Client Hints is worthwhile and are concerned about its implications on the same-origin policy.

**Comment by dmurph** [comment]
I'm also not convinced about using the installed status.

Differential serving of layout is useful, especially because we have two parters that have explicitly said as much, as well as the feature request that this: https://github.com/w3c/manifest/issues/954. So having the display mode CH makes sense here. 

Regarding install status - I don't really see this getting used for quota or run on login in the real world, and I definitely think this would be used for fingerprinting and data collection about the client. I am skeptical if this is even the correct answer for a 'quota' question. I see the cons really outweighing the pros mentioned here. I also don't like how this allows a website to change behavior based on install status. With display mode, all display modes are 'accessible' in an installed app (open-in-tab, fullscreen, or standalone/minimal), so an web app is still incentivized to support them.

I see installation as being something that is more user-focused - the user agent is providing higher-os-level interaction and new features to the user to interact with this web app. They aren't required to install, and they should be able to still use the web app without installing.

**Comment by aarongustafson** [comment]
@marcoscaceres @dmurph While hardly scientific, I found this interesting: https://twitter.com/AaronGustafson/status/1385701549905571840

TLDR; More than 1/3 of developers want to be able to know if their PWA is installed, narrowly beating out a ~30% interest in dark mode colors & icons. Multilingual support was a close third, and knowing the installation source was in last place with ~13% of the vote (sample size: 131 on Twitter)

Trying to get more details on reasons folks want to know about installation. May also run a Kano study on some of this stuff to get a better sense of what developers need from the Manifest, etc.

**Comment by dmurph** [comment]
Cool! I would love to know why they want that signal. Maybe you can create a new thread on why devs want that?

**Comment by marcoscaceres** [comment]
> would love to know why they want that signal. Maybe you can create a new thread on why devs want that?

I'd be keen to dig into that too. If we can look at concrete examples or real apps, that would be great, as we can then figure out what people are trying to do and derive a solution from there.

**Comment by aarongustafson** [comment]
Adding @diekus to this thread since install is something he’s working on as well.
